### PR TITLE
Templates & general improvements

### PR DESCRIPTION
**Pull request by umairidris**
- use pathutil in main
- merge data and project into one recipe, so now calling it deploys a full "data" project
- allow setting multiple vars and outputs in terraform components
- generalize terragrunt dependency component and call it terragrunt 'leaf'
- add new terraform recipe that can create vars, outputs and terragrunt configs from one place
- add terraform and provider versions #13 
- set terraform remote state in helper terragrunt leaf component

### CODE REVIEW

File: templates/tfengine/recipes/project/data.yaml [hunk 1]
```diff
@@ -1,35 +1,36 @@
...
-  data:
-    NAME: "storage_location"
-    TYPE: "string"
-    VALUE: '"{{.STORAGE_LOCATION}}"'
-{{range $i, $_ := .STORAGE_BUCKETS}}
+{{range $i, $_ := get . "STORAGE_BUCKETS"}}
 - name: "bucket-{{.NAME}}"
   component_path: "../../components/project/storage"
-  output_ref: "dir"
+  output_ref: "data_dir"
```
Comment (MartinPetkov) [inline templates/tfengine/recipes/project/data.yaml:32]: Does this need to be unique? It would be the same output_ref for every bucket in the range.

Comment (umairidris) [reply templates/tfengine/recipes/project/data.yaml:32]: No, this tells the engine where to write all the buckets. In this case they are all in the same data deployment.

File: examples/tfengine/full.yaml [hunk 2]
```diff
@@ -61,6 +61,18 @@ templates:
...
     ORG_POLICIES: {} # Use default org policies.
+    CICD:
+      PROJECT_ID: "example-devops"
+      STATE_BUCKET: "example-state-bucket"
+      REPO_OWNER: "GoogleCloudPlatform"
+      REPO_NAME: "example"
+      BRANCH_REGEX: "master"
+      CONTINUOUS_DEPLOYMENT_ENABLED: true
+      TRIGGER_ENABLED: true
+      REJECT_PLAN_DELETIONS: false
```
Comment (xingao267) [inline examples/tfengine/full.yaml:72]: this does not apply anymore

File: templates/tfengine/components/terraform/variables/variables.tf [hunk 3]
```diff
@@ -1,6 +1,8 @@
+{{range .VARS}}
```
Comment (xingao267) [inline templates/tfengine/components/terraform/variables/variables.tf:1]: add - in {{}}?

Comment (umairidris) [reply templates/tfengine/components/terraform/variables/variables.tf:1]: Adding - actually puts no blank lines so I left this in

File: examples/tfengine/full.yaml (hunk 2 above)
Comment (umairidris) [reply examples/tfengine/full.yaml]: Fixed

**Comment by xingao267** [comment]
@googlebot rescan

**Pull request merged**